distribution/lib/Standard/Base/0.0.0-dev/src/Data/Locale.enso

```diff
@@ -18,6 +21,13 @@ polyglot java import java.util.Locale as JavaLocale
    - A country code, which is optional.
    - A variant, which is optional.
 type Locale
+    ## PRIVATE
+       Gets the default drop down option for this encoding.
+    default_widget : Single_Choice
+    default_widget =
+        values = [Option "default" "Locale.default", Option "bangladesh" "Locale.bangladesh", Option "brazil" "Locale.brazil", Option "canada_english" "Locale.canada_english", Option "canada_french" "Locale.canada_french", Option "china" "Locale.china", Option "france" "Locale.france", Option "germany" "Locale.germany", Option "india_hindi" "Locale.india_hindi", Option "india_english" "Locale.india_english", Option "indonesia" "Locale.indonesia", Option "italy" "Locale.italy", Option "japan" "Locale.japan", Option "mexico" "Locale.mexico", Option "nigeria" "Locale.nigeria", Option "pakistan_urdu" "Locale.pakistan_urdu", Option "pakistan_english" "Locale.pakistan_english", Option "poland" "Locale.poland", Option "russia" "Locale.russia", Option "south_korea" "Locale.south_korea", Option "uk" "Locale.uk", Option "us" "Locale.us"]
```

Comment:
I'm wondering if we could get these automatically by relfection? That list is pretty long.

Also I'm wondering if we should make labels more human friendly - capitalized and replacing _ with space.

Reply:
To use reflection, we'd have to annotate the members of the class that we want to show (so we don't show things like .to_text). Is it appropriate to do something like this?

```
    @option_for_display
    pakistan_urdu : Locale
    pakistan_urdu = Locale.from_language_tag "ur-PK"
```

From the tests for Meta.get_annotation, it looks like annotations should always be named after a method's parameters. Is that true? Or is it okay to create annotations with any name?

Reply:
Let's not corrupt the annotation system like that for now.

I think it is fine to hardcode this for now. We can look to improve it an make it dynamically generated at a later point.
Title casing and replacing the `_` would be nice.

Locale.methods - Any.methods would probably be about what we want.

Reply:
Using it as an excuse to test today's build...

![image](https://github.com/enso-org/enso/assets/4699705/b1e18858-ccdb-44bf-b490-f9607d1bc0f4)

```haskell
    operator2 = Meta.meta Locale
    operator21 = operator2.value ...
    operator211 = Meta.type_of operator21
    operator3 = Meta.Type.Value operator211
    operator5 = operator3.methods
    operator6 = operator2.methods
    operator1 = operator5.filter (Filter_Condition.Not_In operator6 )
    operator4 = operator1.filter (Filter_Condition.Not_In ["Value", "new", "default", "from_language_tag"])
    operator7 = operator4.map n-> (n.replace '_' ' ').to_case (if n.length==2 then Case.Upper else Case.Title)
    operator9 = operator4.map n-> "Locale."+n
    operator10 = operator7.zip operator9
    [["Default", "Locale.default"]] + operator10 + [["From Language Tag", "(Locale.from_language_tag 'en-US')"]]
```

Produces a reasonable list:
![image](https://github.com/enso-org/enso/assets/4699705/55ebde59-a80e-48d6-b658-d9f0783f23d4)

Reply:
I was reluctant to generate this by subtracting things -- it means any time someone adds a static member to this class, they have to know to add it to the "remove us" list, if I'm reading this right.

Reply:
Yes, we would need to "black-list" any statics. I don't think a big issue here.

Reply:
A bit hacky, but we can get a unit test that checks all widgets, tries to instantiate them (using reflection like `Polyglot.invoke_method`) and verifies that they in fact return an instance of `Locale`. So if we ever get a 'false positive' method here, it should be caught by such a unit test.

It sounds like possibly an overkill here, so complicated machinery for such a simple thing - so I'm not sure if we really want/need to implement it - but it could be done this way.

Surely that offers best extensibilty - if someone adds a new method, the unit test will catch it and they will know what to amend (the unit test can have a comment with explanation what needs to be done) and if someone adds a new Locale constant - it will get handled automatically. Nice. But indeed a lot of engineering for sth simple 🙂

Reply:
Rather than upper-casing the field name, I'd like to call JavaLocale.display_country -- but to do that, I'd need to be able to get the static field by name. Not sure how to do that:
```
> Polyglot.invoke Locale "poland" []
Evaluation failed with: Message not supported.
java.lang.Exception: Message not supported.
```

Reply:
I'll report this as a bug and use the by-hand capitalization.